File: perf/benchmarks/apps/matrix_multiply/matrix_multiply.go
// Matrix multiplication — naive O(N^3) double-precision matmul (Go).
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"time"
)

func main() {
	n := 250
	if len(os.Args) > 1 {
		v, _ := strconv.Atoi(os.Args[1])
		n = v
	}
	a := make([]float64, n*n)
	b := make([]float64, n*n)
	c := make([]float64, n*n)
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			a[i*n+j] = float64((i*7+j*13)%100) * 0.01
			b[i*n+j] = float64((i*11+j*17)%100) * 0.01
		}
	}
	t0 := time.Now()
	for i := 0; i < n; i++ {
		for k := 0; k < n; k++ {
			aik := a[i*n+k]
			bk := b[k*n : k*n+n]
			ci := c[i*n : i*n+n]
			for j := 0; j < n; j++ {
				ci[j] += aik * bk[j]
			}
		}
	}
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	total := 0.0
	for i := 0; i < n; i++ {
		total += c[i*n+i]
	}
	cs := int64(math.Round(total * 1e6))
	fmt.Printf("CHECKSUM:%d\n", cs)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/apps/matrix_multiply/matrix_multiply.rs
// Matrix multiplication — naive O(N^3) double-precision matmul (Rust).
use std::env;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: usize = if args.len() > 1 { args[1].parse().unwrap_or(250) } else { 250 };
    let mut a: Vec<f64> = vec![0.0; n * n];
    let mut b: Vec<f64> = vec![0.0; n * n];
    let mut c: Vec<f64> = vec![0.0; n * n];
    for i in 0..n {
        for j in 0..n {
            a[i * n + j] = ((i as i64 * 7 + j as i64 * 13) % 100) as f64 * 0.01;
            b[i * n + j] = ((i as i64 * 11 + j as i64 * 17) % 100) as f64 * 0.01;
        }
    }
    let t0 = Instant::now();
    for i in 0..n {
        for k in 0..n {
            let aik = a[i * n + k];
            let (bk_off, ci_off) = (k * n, i * n);
            for j in 0..n {
                c[ci_off + j] += aik * b[bk_off + j];
            }
        }
    }
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    let mut total = 0.0;
    for i in 0..n {
        total += c[i * n + i];
    }
    let cs = (total * 1e6).round() as i64;
    println!("CHECKSUM:{}", cs);
    println!("ELAPSED_MS:{:.6}", ms);
}
